File: horizons_py/horizons/optimization.py
from __future__ import annotations

from typing import Any, Dict, List, Optional
from uuid import UUID

from .client import HorizonsClient
from . import models


class OptimizationAPI:
    def __init__(self, client: HorizonsClient) -> None:
        self._client = client

    async def run(
        self,
        *,
        cfg: Dict[str, Any],
        initial_policy: Dict[str, Any],
        dataset: Dict[str, Any],
        project_id: Optional[UUID] = None,
    ) -> models.OptimizationRunRow:
        body: Dict[str, Any] = {"cfg": cfg, "initial_policy": initial_policy, "dataset": dataset}
        if project_id:
            body["project_id"] = str(project_id)
        resp = await self._client._request("POST", "/api/v1/optimization/run", json=body)
        data = await self._client.json(resp)
        return models.OptimizationRunRow.model_validate(data)

    async def status(
        self, *, project_id: Optional[UUID] = None, limit: int = 50, offset: int = 0
    ) -> List[models.OptimizationRunRow]:
        params: Dict[str, Any] = {"limit": limit, "offset": offset}
        if project_id:
            params["project_id"] = str(project_id)
        resp = await self._client._request("GET", "/api/v1/optimization/status", params=params)
        data = await self._client.json(resp)
        return [models.OptimizationRunRow.model_validate(item) for item in data]

    async def reports(
        self, *, project_id: Optional[UUID] = None, run_id: Optional[UUID] = None, limit: int = 50, offset: int = 0
    ) -> Dict[str, Any]:
        params: Dict[str, Any] = {"limit": limit, "offset": offset}
        if project_id:
            params["project_id"] = str(project_id)
        if run_id:
            params["run_id"] = str(run_id)
        resp = await self._client._request("GET", "/api/v1/optimization/reports", params=params)
        return await self._client.json(resp)

File: horizons_ts/src/optimization.ts
import { HorizonsClient } from "./client";
import { OptimizationRunRow, UUID } from "./types";

export class OptimizationAPI {
  constructor(private client: HorizonsClient) {}

  async run(input: {
    cfg: Record<string, unknown>;
    initial_policy: Record<string, unknown>;
    dataset: Record<string, unknown>;
    project_id?: UUID;
  }): Promise<OptimizationRunRow> {
    const resp = await this.client.request<OptimizationRunRow>("/api/v1/optimization/run", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(input),
    });
    return resp;
  }

  async status(params: { project_id?: UUID; limit?: number; offset?: number } = {}): Promise<OptimizationRunRow[]> {
    const search = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => {
      if (v !== undefined && v !== null) search.append(k, String(v));
    });
    const path = `/api/v1/optimization/status${search.toString() ? `?${search.toString()}` : ""}`;
    return await this.client.request<OptimizationRunRow[]>(path, { method: "GET" });
  }

  async reports(params: { project_id?: UUID; run_id?: UUID; limit?: number; offset?: number } = {}): Promise<Record<string, unknown>> {
    const search = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => {
      if (v !== undefined && v !== null) search.append(k, String(v));
    });
    const path = `/api/v1/optimization/reports${search.toString() ? `?${search.toString()}` : ""}`;
    return await this.client.request<Record<string, unknown>>(path, { method: "GET" });
  }
}
